htsget-search/src/storage/url/mod.rs

```diff
@@ -0,0 +1,1178 @@
+      true,
+    );
+
+    let mut headers = HeaderMap::default();
+    let options = test_range_options(&mut headers);
+
+    assert_eq!(
+      storage
+        .format_url(
+          "assets/key1",
+          options,
+          &Uri::from_str("https://localhost:8080").unwrap()
+        )
+        .await
+        .unwrap(),
+      HtsGetUrl::new("https://localhost:8080/assets/key1")
+        .with_headers(Headers::default().with_header(AUTHORIZATION.as_str(), "secret"))
+    );
+  }
+
+  #[tokio::test]
+  async fn format_url_different_response_scheme() {
+    let storage = UrlStorage::new(
+      test_client(),
+      endpoints_test(),
+      Uri::from_str("http://example.com").unwrap(),
+      true,
+    );
+
+    let mut headers = HeaderMap::default();
+    let options = test_range_options(&mut headers);
+
+    assert_eq!(
+      storage
+        .format_url(
+          "assets/key1",
+          options,
+          &Uri::from_str("http://example.com").unwrap()
+        )
+        .await
+        .unwrap(),
+      HtsGetUrl::new("http://example.com/assets/key1")
+        .with_headers(Headers::default().with_header(AUTHORIZATION.as_str(), "secret"))
+    );
+  }
+
+  #[tokio::test]
+  async fn format_url_no_headers() {
+    let storage = UrlStorage::new(
+      test_client(),
+      endpoints_test(),
+      Uri::from_str("https://localhost:8081").unwrap(),
+      false,
+    );
+
+    let mut headers = HeaderMap::default();
+    let options = test_range_options(&mut headers);
+
+    assert_eq!(
+      storage.range_url("assets/key1", options,).await.unwrap(),
+      HtsGetUrl::new("https://localhost:8081/assets/key1")
+    );
+  }
+
+  #[tokio::test]
+  async fn test_endpoints_with_real_file() {
+    with_url_test_server(|url| async move {
+      let storage = UrlStorage::new(
+        test_client(),
+        endpoints_from_url_with_path(&url),
+        Uri::from_str("http://example.com").unwrap(),
+        true,
+      );
+
+      let mut header_map = HeaderMap::default();
+      test_headers(&mut header_map);
+      let request =
+        HtsgetRequest::new_with_id("htsnexus_test_NA12878".to_string()).with_headers(header_map);
+      let query = Query::new(
+        "htsnexus_test_NA12878",
+        Format::Bam,
+        request,
+        Default::default(),
+      )
+      .with_reference_name("11")
+      .with_start(5015000)
+      .with_end(5050000);
+
+      let searcher = HtsGetFromStorage::new(storage);
+      let response = searcher.search(query.clone()).await;
+
+      let expected_response = Ok(expected_bam_response());
+      assert_eq!(response, expected_response);
+
+      let (bytes, _) = test_bam_file_byte_ranges(
+        response.unwrap(),
+        default_dir().join("data/bam/htsnexus_test_NA12878.bam"),
+      )
+      .await;
+
+      parse_as_bgzf(bytes).await;
+    })
+    .await;
+  }
+
+  #[cfg(feature = "crypt4gh")]
+  #[tokio::test]
+  async fn test_endpoints_with_real_file_encrypted() {
```

Comment:
This is the primary test that has encrypted data going between the mock client, htsget-rs, and the mock `UrlStorage` backend.